lib/chef/knife/ec_error_handler.rb

```diff
@@ -0,0 +1,150 @@
+  class Knife
+    # This class handles the errors that we might encounter on a
+    # backup or restore, it stors the errors inside a specific file
+    # inside the working directory.
+    class EcErrorHandler
+
+      # Creates a new instance of the EcErrorHandler to start
+      # adding errors during a backup or restore.
+      def initialize(working_dir, process)
+        @err_dir = File.join(working_dir, 'errors')
+        FileUtils.mkdir_p(@err_dir)
+
+        @err_file = case process
+          when Chef::Knife::EcRestore
+            File.join(@err_dir, "restore-#{Time.now.iso8601}.json")
+          when Chef::Knife::EcBackup
+            File.join(@err_dir, "backup-#{Time.now.iso8601}.json")
+          else
+            File.join(@err_dir, "other-#{Time.now.iso8601}.json")
+        end
+
+        # exit handler
+        at_exit { EcErrorHandler.display(@err_file) }
+      end
+
+      # Add an exception to the error file.
+      #
+      # For now we are writing all the errors to a single file, but
+      # in the future we would like to be able to generate a full path
+      # just as we do when we are backing up the Server, just that putting
+      # the file inside `{work_dir}/errors/{path}`
+      #
+      # Example:
+      # A failed user
+      # => {work_dir}/errors/users/afiune.json
+      # A failed cookbook
+      # => {work_dir}/errors/cookbooks/burger.json
+      # A failed environment
+      # => {work_dir}/errors/environment/dev.json
+      #
+      # The advantages of this schema is the ability to retry the backup or
+      # restore and pick up where we left.
+      def add(ex)
+        msg = {
+          timestamp:  Time.now,
+          message:    ex.message,
+          backtrace:  ex.backtrace
+        }
+        if ex.respond_to?(:knife_ec_backup_rest_request=)
+          msg.merge!(ex.knife_ec_backup_rest_request)
+        end
+        EcErrorHandler.lock_file(@err_file, 'a') do |f|
+          f.write(Chef::JSONCompat.to_json_pretty(msg))
+        end
+      end
+
+      def self.lock_file(file_name, mode)
+        File.open(file_name, mode) do |f|
+          begin
+            f.flock File::LOCK_EX
+            yield f
+          ensure
+            f.flock File::LOCK_UN
+          end
+        end
+      end
+
+      def self.display(file_name)
+        puts "\nError Summary Report"
+        EcErrorHandler.lock_file(file_name, 'r') do |f|
+          f.each_line do |line|
+            puts line
+          end
+        end
+        puts "\nError(s) Summary file located at: '#{file_name}'"
+      end
+    end
+  end
+end
+
+# Openning the ServerAPI class to inject the HTTP request context into
+# any `Net::HTTPServerException` exception, overriding the get, put, post
+# and delete methods, incorporating a `rescue` block in each, adding the
+# extra context from instance variables as well as the request path to the
+# object, then bubbling the error up in the stack with raise.
+Chef::ServerAPI.class_eval do
```

Comment:
Uhhhh a shiny monkey is looking at me! Thanks @lamont-granquist @danielsdeleo I think we can leverage that!!! 🐒 

On a side note, we should consider have a conversation around extending ChefFs to be flexible enough to do this kind of exception catching, something like what we did with `:chef_rest_request` in core of chef. The reason why is just that we would like to catch any failures on the backups and store them on a single location. I'll be happy to talk further more about this. 😄